text/0240-Sampling_Configuration.md

```diff
@@ -0,0 +1,232 @@
+
+## Motivation
+
+The need for sampling configuration has been explicitly or implicitly indicated in several discussions, some of them going back a number of years, see for example
+
+- issue [173](https://github.com/open-telemetry/opentelemetry-specification/issues/173): Way to ignore healthcheck traces when using automatic tracer across all languages?
+- issue [679](https://github.com/open-telemetry/opentelemetry-specification/issues/679): Configuring a sampler from an environment variable
+- issue [1060](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1060): Exclude URLs from Tracing
+- issue [1844](https://github.com/open-telemetry/opentelemetry-specification/issues/1844): Composite Sampler
+- issue [2085](https://github.com/open-telemetry/opentelemetry-specification/issues/2085): Remote controlled sampling rules using attributes
+- issue [3205](https://github.com/open-telemetry/opentelemetry-specification/issues/3205): Equivalent of "logLevel" for spans
+- discussion [3725](https://github.com/open-telemetry/opentelemetry-specification/discussions/3725): Overriding Decisions of the Built-in Trace Samplers via a "sample.priority" Like Span Attribute
+
+A number of custom samplers are already available as independent contributions
+([RuleBasedRoutingSampler](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/sampler/RuleBasedRoutingSampler.java),
+[LinksBasedSampler](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/sampler/LinksBasedSampler.java),
+and, of course, the latest and greatest [Consistent Probability Sampling](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/consistent-sampling)) or just as ideas. They all share the same pain point, which is lack of easy to use configuration mechanism.
+
+Even when the code for these samplers is available, their use is not very simple. In case of Java, they require writing a custom agent [extension](https://opentelemetry.io/docs/instrumentation/java/automatic/extensions/). This can become a hurdle, especially for the Open Telemetry users with no hands-on coding experience.
+
+## The Goal
+
+The goal of this proposal is to create an open-ended configuration schema which supports not only the current set of SDK standard samplers, but also non-standard ones, and even samplers that will be added in the future. Furthermore the samplers should be composable together, as it is often required by the users.
+
+In contrast, the existing configuration schemas, such as [Jaeger sampling](https://www.jaegertracing.io/docs/1.50/sampling/#file-based-sampling-configuration), or Agent [OTEP 225 - Configuration proposal](https://github.com/open-telemetry/oteps/pull/225) address sampling configuration with a limited known set of samplers only.
```

Comment:
> Furthermore the samplers should be composable together,

File configuration supports this.

File based configuration will support custom extension samplers. One of the stated goals is:

> Custom span processors, exporters, samplers, or other user defined extension components can be configured using this format.

There's an outstanding question of how exactly these extension components are called with configuration such that an instance can be configured. My working mental model is that we should have a generic configuration properties class which has accessors for reading configuration properties. Then, when file configuration comes across a reference to an extension component, it tries to load and obtain (language specific how this will work) an instance of it by passing the the generic configuration properties to it. The component will then be responsible for reading any of its fields from configuration properties and providing a configured instance.

With that said, I think we should extend the specification to include additional built-in samplers to accommodate common use cases. Having these would imply having dedicated types defined in the [opentelemetry-configuration](https://github.com/open-telemetry/opentelemetry-configuration) schema.

Reply:
Yes, I had a look at the file based configuration and I saw the sampling configuration there, and it looks like it takes only samplers from a predefined set. 
If I understand you correctly, extending it to cover any (or almost any) sampler would require a 2-step process. Custom samplers would have to get their types and parameters declared there first. Do you expect the Open Telemetry end users being able to do that on their own?

Reply:
I'm not referring to jaeger file based configuration. I'm referring to the prototype OpenTelemetry file configuration specification, which was proposed in this [otep](https://github.com/open-telemetry/oteps/pull/225), and which is specified in this [document](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md), and includes a schema defined in [opentelemetry-configuration](https://github.com/open-telemetry/opentelemetry-configuration).

[Here's](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/examples/kitchen-sink.yaml#L343-L349) some example config where we specify that the sampler is the parent based sampler, and go on to specify samplers as inputs to the parent based sampler's configuration. In this situation, the parent based sampler delegates to the trace id ratio sampler when a certain condition is true (i.e. the span has no parent), but the file configuration schema is designed to be able to accommodate arbitrarily complex configurations and support custom extension components like samplers. This means that if you can dream up how to specify it in YAML, you should be able to use file configuration to accomplish it. 

Here's a "[steel thread](https://github.com/open-telemetry/opentelemetry-java-examples/pull/227)" I've been working through to prove out the concepts, which uses file configuration to configure a custom sampler (one that is published form `opentelemetry-java-contrib`) to filter out spans with matching attribute conditions.

However, things are easier when you can reference components which are part of the specification and built into SDK implementations. And I think it would be good to spec out a rule based sampler, with a declarative syntax for specifying conditions and samplers that should be used when those conditions are met.

Reply:
Got it. Apologies for my misunderstanding! As someone relatively out-of-the-loop, the idea of embedding this configuration within a broader SDK schema seems very sensible to me.

Reply:
@jack-berg it looks like your work in https://github.com/open-telemetry/opentelemetry-specification/pull/3802 supports this OTEP nicely. Thank you!